README.md
# yolo_v8_segmentation

::: {.cell 0=‘h’ 1=‘i’ 2=‘d’ 3=‘e’ execution_count=4}

``` python
from yolo_v8_segmentation.core import *
from yolo_v8_segmentation.data_preparation import *
from pathlib import Path
from fastcore.all import * 
from nbdev.showdoc import show_doc
```

:::

> segementation training and inference with yolov8 model

This file will become your README and also the index of your
documentation.

## Install

``` sh
pip install yolo_v8_segmentation
```

## How to use

### Visualize mask and image

``` python
from yolo_v8_segmentation.viz_utils import *
```

``` python
msk_path = Path(r'/home/hasan/Schreibtisch/projects/data/microscopy/patch_train_masks')
img_path = Path(r'/home/hasan/Schreibtisch/projects/data/microscopy/patch_train_images/')
```

``` python
idx = 2
image = img_path.ls()[idx]
msk = msk_path.ls()[idx]
```

``` python
show_doc(overlay_mask_border_on_image)
```

------------------------------------------------------------------------

<a
href="https://github.com/HasanGoni/yolo_v8_segmentation/blob/main/yolo_v8_segmentation/viz_utils.py#LNone"
target="_blank" style="float:right; font-size:smaller">source</a>

### overlay_mask_border_on_image

>      overlay_mask_border_on_image (im_path, msk_path, border_color=(0, 1, 0),
>                                    border_width=1)

Overlays the border of a binary mask on a grayscale image and displays
the result using matplotlib.

Args: image (numpy.ndarray): Grayscale image. mask (numpy.ndarray):
Binary mask of the same size as the image. border_color (tuple): RGB
color for the mask border in the range \[0, 1\]. border_width (int):
Width of the border.

Returns: None: The function displays a plot.

``` python
overlay_mask_border_on_image(im_path=image, msk_path=msk)
```

![](index_files/figure-commonmark/cell-7-output-1.png)

``` python
overlay_mask(im_path=image, msk_path=msk)   
```

![](index_files/figure-commonmark/cell-8-output-1.png)

## Yolo annotation visualization

- Upto now it is just a folder where there are some image and mask
  pairs. Previous method shows the masks and images
- In case data is prepared in `.txt` format for training yolo, it is
  important to check it whether the annotations are correct or not. This
  function will help to visualize the annotations on the image.

``` python
yolo_mask_path = Path(r'/home/hasan/Schreibtisch/projects/data/microscopy/yolo_dataset_train')
im_path = Path(r'/home/hasan/Schreibtisch/projects/data/microscopy/patch_train_images/')
```

``` python
overlay_yolo_mask(im_path=im_path, msk_path=yolo_mask_path)
```

![](index_files/figure-commonmark/cell-10-output-1.png)

- in case specific image wanted to see, then one need to pass the image
  name to the function
  - here I am using im_name=None, so random image is selected

``` python

overlay_yolo_mask(im_path=im_path, msk_path=yolo_mask_path, im_name=None, fill=True, alpha=0.3)
```

![](index_files/figure-commonmark/cell-11-output-1.png)

``` python

overlay_yolo_mask(
    im_path=im_path, 
    msk_path=yolo_mask_path, 
    im_name=None, 
    color='red',
    fill=True, 
    alpha=0.3)
```

![](index_files/figure-commonmark/cell-12-output-1.png)

## Convert label to yolo format

### First coco format json file

``` python
- Normaly there are some labelling tool where you can label your images and then upload them direct in the yolo format.
- But we are supossing that, you have some binary masks and then you want to convert them to yolo format.
- First we will convert to coco format and then to yolo format.
```

``` python
# in my case I have only one class and others are background, in case of more class, one should add name and id for the masks

category_ids = {
    "object": 1,
}
```

``` python
# where masks(in my case binary masks) are stored
trn_msk_path = Path(r'/home/hasan/Schreibtisch/projects/data/microscopy/patch_train_masks')
# this will be json path for the coco format
trn_json_path = Path(r'/home/hasan/Schreibtisch/projects/data/microscopy/patch_train_masks/train.json')
process_masks(
    mask_path=trn_msk_path, 
    json_path=trn_json_path, 
    category_ids=category_ids)
```

      0%|          | 0/1642 [00:00<?, ?it/s]

### Now coco foramt

``` python
create_yolo_dataset(
    img_path=trn_img_path,# train images path 
    output_path=trn_output_path, # output for each image normally some txt file
    json_path=trn_json_path, # json path for coco format
)
```

``` python
# for yolo to run one needs to create a yaml file to tell where the data is situated
# coco expect following format
# train: /path/to/train/images
# label: /path/to/train/labels
# valid: /path/to/valid/images
# valid_label: /path/to/valid/labels

create_yaml(
    json_path=trn_json_path, # coco format json path
    yaml_path=f'{path}/data.yaml', # yaml path for saving
    train_path=yolo_trn_images,  # where train images are stored
    val_path=trn_img_path, # where validaiton images are stored
    )
```


CHANGELOG.md
# Release notes

<!-- do not remove -->

## 0.0.1


### Bugs Squashed

- just testing ([#1](https://github.com/HasanGoni/yolo_v8_segmentation/issues/1))


yolo_v8_segmentation/_modidx.py
# Autogenerated by nbdev

d = { 'settings': { 'branch': 'main',
                'doc_baseurl': '/yolo_v8_segmentation',
                'doc_host': 'https://HasanGoni.github.io',
                'git_url': 'https://github.com/HasanGoni/yolo_v8_segmentation',
                'lib_path': 'yolo_v8_segmentation'},
  'syms': { 'yolo_v8_segmentation.core': {'yolo_v8_segmentation.core.foo': ('core.html#foo', 'yolo_v8_segmentation/core.py')},
            'yolo_v8_segmentation.data_preparation': { 'yolo_v8_segmentation.data_preparation.create_yaml': ( 'data_preparation.html#create_yaml',
                                                                                                              'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.create_yolo_dataset': ( 'data_preparation.html#create_yolo_dataset',
                                                                                                                      'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.from_contr_to_annotation': ( 'data_preparation.html#from_contr_to_annotation',
                                                                                                                           'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.get_annotations': ( 'data_preparation.html#get_annotations',
                                                                                                                  'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.get_contours': ( 'data_preparation.html#get_contours',
                                                                                                               'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.get_file_info': ( 'data_preparation.html#get_file_info',
                                                                                                                'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.get_mask_info': ( 'data_preparation.html#get_mask_info',
                                                                                                                'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.normalized_polygon': ( 'data_preparation.html#normalized_polygon',
                                                                                                                     'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.process_masks': ( 'data_preparation.html#process_masks',
                                                                                                                'yolo_v8_segmentation/data_preparation.py'),
                                                       'yolo_v8_segmentation.data_preparation.read_json': ( 'data_preparation.html#read_json',
                                                                                                            'yolo_v8_segmentation/data_preparation.py')},
            'yolo_v8_segmentation.viz_utils': { 'yolo_v8_segmentation.viz_utils.overlay_mask': ( 'viz_utils.html#overlay_mask',
                                                                                                 'yolo_v8_segmentation/viz_utils.py'),
                                                'yolo_v8_segmentation.viz_utils.overlay_mask_border_on_image': ( 'viz_utils.html#overlay_mask_border_on_image',
                                                                                                                 'yolo_v8_segmentation/viz_utils.py'),
                                                'yolo_v8_segmentation.viz_utils.overlay_yolo_mask': ( 'viz_utils.html#overlay_yolo_mask',
                                                                                                      'yolo_v8_segmentation/viz_utils.py')},
            'yolo_v8_segmentation.yolo_v8_training': {}}}

yolo_v8_segmentation/viz_utils.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/02_viz_utils.ipynb.

# %% auto 0
__all__ = ['get_name', 'overlay_mask', 'overlay_mask_border_on_image', 'overlay_yolo_mask']

# %% ../nbs/02_viz_utils.ipynb 4
import cv2
from pathlib import Path
import numpy as np
from fastcore.all import *
import matplotlib.pyplot as plt
import matplotlib as mpl
from scipy.ndimage import binary_dilation, binary_erosion
import matplotlib.patches as patches
import matplotlib as mpl

# %% ../nbs/02_viz_utils.ipynb 6
get_name = np.vectorize(lambda x: Path(x).name)

# %% ../nbs/02_viz_utils.ipynb 13
def overlay_mask(
        im_path, 
        msk_path, 
        overlay_clr=(0, 1, 0),
        scale=1,
        alpha=0.5,
        ):
    'Creaete a overlay image from image and mask'
    # Read the grayscale image
    gray_img = cv2.imread(f'{im_path}', cv2.IMREAD_GRAYSCALE)
    if gray_img is None:
        raise ValueError("Could not read the grayscale image")

    # Read the mask image
    mask_img = cv2.imread(f'{msk_path}', cv2.IMREAD_GRAYSCALE)
    mask_img = mask_img.astype(bool)
    if mask_img is None:
        raise ValueError("Could not read the mask image")

    # Check if dimensions of both images are the same
    if gray_img.shape != mask_img.shape:
        raise ValueError("Dimensions of grayscale image and mask do not match")

    # Convert image to 3 channels
    rgb_img = np.stack([gray_img]*3, axis=-1)/255
    fig, ax = plt.subplots()
    ax.imshow(rgb_img)

    clrd_overlay = np.zeros_like(rgb_img)
    clrd_overlay[mask_img]=overlay_clr
    ax.imshow(clrd_overlay, alpha=alpha)





# %% ../nbs/02_viz_utils.ipynb 15
def overlay_mask_border_on_image(im_path, msk_path, border_color=(0, 1, 0), border_width=1):
    """
    Overlays the border of a binary mask on a grayscale image and displays the result using matplotlib.
    
    Args:
    image (numpy.ndarray): Grayscale image.
    mask (numpy.ndarray): Binary mask of the same size as the image.
    border_color (tuple): RGB color for the mask border in the range [0, 1].
    border_width (int): Width of the border.
    
    Returns:
    None: The function displays a plot.
    """
    gray_img = cv2.imread(f'{im_path}', cv2.IMREAD_GRAYSCALE)
    if gray_img is None:
        raise ValueError("Could not read the grayscale image")

    # Read the mask image
    mask_img = cv2.imread(f'{msk_path}', cv2.IMREAD_GRAYSCALE)
    mask_img = mask_img.astype(bool)
    if mask_img is None:
        raise ValueError("Could not read the mask image")

    # Check if dimensions of both images are the same
    if gray_img.shape != mask_img.shape:
        raise ValueError("Dimensions of grayscale image and mask do not match")
    # Ensure the mask is boolean

    # Find the borders of the mask
    dilated_mask = binary_dilation(mask_img, iterations=border_width)
    eroded_mask = binary_erosion(mask_img, iterations=border_width)
    border = dilated_mask & ~eroded_mask

    # Convert grayscale image to RGB
    rgb_image = np.stack([gray_img]*3, axis=-1) / 255.0  # Normalize for matplotlib

    # Apply the colored border
    rgb_image[border] = border_color

    # Display the image using matplotlib
    plt.imshow(rgb_image, cmap='gray')
    plt.axis('off')  # Turn off axis numbers
    plt.show()



# %% ../nbs/02_viz_utils.ipynb 20
def overlay_yolo_mask(
        im_path, 
        msk_path, 
        color=None, # color values other matplotlib colors
        idx=None, # Index of image in the folder other random will be chosen
        im_name=None, # Imge name with extension
        fill=False, # whther to fill the mask or not
        alpha=0.3 # alpha value for the mask for transparency
        ):
    'Creaete a overlay image from image and mask'


    im_idx = idx if idx is not None else np.random.choice(len(im_path.ls()), 1)[0]
    im_name = im_name if im_name is not None else im_path.ls()[im_idx].name
    mask_name = f'{msk_path}/{im_name}'.replace('png', 'txt')



    # Read the grayscale image
    gray_img = cv2.imread(f'{im_path}/{im_name}', cv2.IMREAD_GRAYSCALE)
    if gray_img is None:
        raise ValueError("Could not read the grayscale image")
    if color is None:
        colors = mpl.colormaps.get_cmap('tab10')



    # Convert image to 3 channels
    rgb_img = np.stack([gray_img]*3, axis=-1)

    image_h, image_w, _ = rgb_img.shape
    fig, ax = plt.subplots()
    ax.imshow(rgb_img)

    with open(f'{mask_name}', 'r') as f:
        for line in f:
            parts = line.strip().split()
            category_id = int(parts[0])
            color = colors(category_id  %10) if color is None else color

            polygon = [float(coord) for coord in parts[1:]]
            polygon = np.array(polygon).reshape(-1, 2) * np.array([image_w, image_h])
            patch = patches.Polygon(polygon, closed=True, edgecolor=color, fill=fill, alpha=alpha, linewidth=3)
            ax.add_patch(patch)
    plt.axis('off')
    plt.title('Image with mask')
    plt.show()


yolo_v8_segmentation/data_preparation.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/01_data_preparation.ipynb.

# %% auto 0
__all__ = ['get_name', 'get_contours', 'from_contr_to_annotation', 'get_mask_info', 'process_masks', 'read_json', 'get_file_info',
           'get_annotations', 'normalized_polygon', 'create_yolo_dataset', 'create_yaml']

# %% ../nbs/01_data_preparation.ipynb 1
from pathlib import Path
from fastcore.all import *
import cv2
import numpy as np
from typing import List, Tuple, Union, Callable, Optional, Dict
from tqdm.auto import tqdm
import matplotlib.pyplot as plt 
import json
import yaml

# %% ../nbs/01_data_preparation.ipynb 9
get_name = np.vectorize(lambda x: Path(x).name)

# %% ../nbs/01_data_preparation.ipynb 10
def get_contours(img:np.ndarray):
    'get contours from masks'

    _, thresh = cv2.threshold(img, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
    return cv2.findContours(thresh, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)[0]


# %% ../nbs/01_data_preparation.ipynb 11
def from_contr_to_annotation(
                            sn_cntr:list, # single contour
                            consider_min_area:bool=True,# whether to use min_area parameter
                            min_area:int=0,
                            )->Tuple:
    'Create annotation dict from  a single contour'
    bbox = cv2.boundingRect(sn_cntr)
    area = cv2.contourArea(sn_cntr)
    segmentation = sn_cntr.flatten().tolist()
    if consider_min_area:
        if area > min_area:
            return bbox, area, segmentation
        return None, None, None
    else:
        return bbox, area, segmentation



# %% ../nbs/01_data_preparation.ipynb 12
def get_mask_info(
        msk_path, 
        min_area=0,
        )->Tuple:
    all_masks = msk_path.ls()

    image_infos = []
    annotations = []
    annotation_id=0
    for idx, msk_fn in tqdm(enumerate(all_masks),total=len(all_masks)):
        image_id = idx +1
        file_name = msk_fn.name

        mask = cv2.imread(str(msk_fn), cv2.IMREAD_GRAYSCALE)
        height, width = mask.shape

        if file_name not in map(itemgetter('file_name'), image_infos):  
            image_info = {
                'id': image_id,
                'width': width,
                'height': height,
                'file_name': file_name}
            image_infos.append(image_info)
        else:
            image_info = list(filter(lambda x: x['file_name'] == file_name, image_infos))[0]
        
        cntrs = get_contours(mask)
        for cntr in cntrs:
            bbox, area, segmentation = from_contr_to_annotation(cntr, min_area=min_area)
            if bbox:
                annotation = {
                    'image_id': image_id,
                    'id': annotation_id,
                    'category_id': 1,
                    'iscrowd': 0,
                    'area': area,
                    'bbox': bbox,
                    'segmentation': [segmentation]
                }
                annotations.append(annotation)
                annotation_id +=1
    return image_infos, annotations, annotation_id


        

        



# %% ../nbs/01_data_preparation.ipynb 15
def process_masks(
        mask_path:Union[str, Path],
        json_path:Union[str, Path],
        category_ids:Dict,
        ):
    coco_format = {
        "info": {},
        "licenses": [],
        "images":[],
        "categories": [{"id":v, "name":k, "supercategory":k } for k, v in category_ids.items()],
        "annotations":[]

    }

    coco_format['images'], coco_format['annotations'], ann_cnt = get_mask_info(mask_path)
    with open(json_path, 'w') as f:
        json.dump(coco_format, f, sort_keys=True, indent=4)

# %% ../nbs/01_data_preparation.ipynb 21
def read_json(file_path):
    with open(file_path, 'r') as f:
        data = json.load(f)
    return data


# %% ../nbs/01_data_preparation.ipynb 26
def get_file_info(json_data:dict, file_name:str):
    return list(filter(lambda x: x['file_name'] == file_name, json_data['images']))[0]

# %% ../nbs/01_data_preparation.ipynb 27
def get_annotations(json_data:dict, file_name:str):
    image_id = get_file_info(json_data, file_name)['id']
    return list(filter(lambda x: x['image_id'] == image_id, json_data['annotations']))

# %% ../nbs/01_data_preparation.ipynb 28
def normalized_polygon(polygon:List, width:int, height:int):
    'normalize polygon coordinates based on image height and width'

    n_p = np.array(polygon).reshape(-1, 2) / np.array([width, height])
    return n_p.flatten().tolist()



# %% ../nbs/01_data_preparation.ipynb 30
def create_yolo_dataset(
    img_path:Union[str, Path],
    output_path:Union[str, Path],
    json_path:Union[str, Path],
    )->None:

    'Create yolo dataset from coco format'

    Path(output_path).mkdir(parents=True, exist_ok=True)

    json_data = read_json(json_path)

    # getting the names of  the images
    image_names = get_name(img_path.ls())

    for i in tqdm(image_names,total=len(image_names)):

        file_info = get_file_info(json_data, i)
        image_height = file_info['height']
        image_width = file_info['width']
        image_annotation = get_annotations(json_data, i)
        # in case annotations available
        if image_annotation:

            # Creating txt file for each image
            with open(output_path/f'{Path(i).stem}.txt', 'w') as f_o:
                for ann in image_annotation:
                    current_cat = ann['category_id'] -1
                    polygon = ann['segmentation'][0]
                    norm_poly = normalized_polygon(
                                                polygon, 
                                                width=image_width, 
                                                height=image_height)

                    f_o.write(f'{current_cat} {" ".join(map(str, norm_poly))}\n')


# %% ../nbs/01_data_preparation.ipynb 34
def create_yaml(
    json_path:Union[str, Path], # json path with its name
    yaml_path:Union[str, Path], # output path with yaml name
    train_path:Union[str, Path],# train images path
    val_path:Union[str, Path], # validation images path
    test_path:Union[str, Path, None]=None,
    )->None:

    ' Create a yaml with trianing and validation images path'


    json_data = read_json(json_path)
    names = [cat['name']for cat in json_data['categories']]

    # Number of classes
    nc = len(names)
    yaml_data ={
        'names': names,
        'nc': nc,
        'test':str(test_path) if test_path else '',
        'train':str(train_path),
         'val':str(val_path)

    }
    with open(yaml_path, 'w') as f:
        yaml.dump(
                yaml_data, 
                f, 
                default_flow_style=False
                )

